### PR TITLE
Add CellCostChanged event to Locomotor.

### PR DESCRIPTION
**Pull request by RoosterDragon**
This event allows subscribers to be made aware when the terrain costs for a cell change.

----

This is motivated by #19591, which will make use of this new event. However it's a general concept that could have other uses as well.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs [hunk 1]
```diff
@@ -431,7 +436,12 @@ void UpdateCellCost(CPos cell)
 
 			var cache = cellsCost[cell.Layer];
 
-			cache[cell] = cost;
+			var uv = cell.ToMPos(world.Map);
+			var oldCost = cache[uv];
+			cache[uv] = cost;
+
+			var cellCostChanged = CellCostChanged;
+			cellCostChanged?.Invoke(cell, oldCost, cost);
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/World/Locomotor.cs]: `CellCostChanged?.Invoke(cell, oldCost, cost);` ?

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs [hunk 2]
```diff
@@ -431,7 +436,11 @@ void UpdateCellCost(CPos cell)
 
 			var cache = cellsCost[cell.Layer];
 
-			cache[cell] = cost;
+			var uv = cell.ToMPos(world.Map);
+			var oldCost = cache[uv];
+			cache[uv] = cost;
+
+			CellCostChanged?.Invoke(cell, oldCost, cost);
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/World/Locomotor.cs]: Would there be any performance benefits to check if CellCostChanged is null and then dont access the oldCost?

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/World/Locomotor.cs]: A minor benefit. Have updated the code to account for that.

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs [hunk 3]
```diff
@@ -432,8 +437,17 @@ void UpdateCellCost(CPos cell)
 				cost = terrainInfos[index].Cost;
 
 			var cache = cellsCost[cell.Layer];
-
-			cache[cell] = cost;
+			if (CellCostChanged is null)
+			{
+				cache[cell] = cost;
+			}
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/World/Locomotor.cs]:
```suggestion
			if (CellCostChanged is null)
				cache[cell] = cost;
```

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/World/Locomotor.cs]: Updated

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs [hunk 4]
```diff
@@ -432,8 +437,15 @@ void UpdateCellCost(CPos cell)
 				cost = terrainInfos[index].Cost;
 
 			var cache = cellsCost[cell.Layer];
-
-			cache[cell] = cost;
+			if (CellCostChanged is null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/Locomotor.cs]: I don't think we're using `is` anywhere else in the codebase, so `== null` for consistency?

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/World/Locomotor.cs]: Updated

**Review by pchote — APPROVED** [review]

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/30571a4c08ebf08198a6f66b59690637a06228f4...52942d3ce2aab4b6c59614746a0cac2a13c939e9

**Pull request merged**